### PR TITLE
API Isolated and NonIsolated CPU changes

### PR DESCRIPTION
**Pull request by davidvossel**
## Overview

The spec.CPU.NonIsolated option is redundant. It can be represented by the same cpu set that spec.CPU.Reserved represents now. If we find we need fine granular control over how the CPUs in the reserved set are used, we can re-approach making an addition to the API at that point. 

The spec.CPU.Isolated option needs two modes, one that just uses the tuned options for cpu partitioning, and another that includes both the tuned settings plus setting the isolcpu kernel option.

The reasoning here is that using the isolcpus kernel option is not always appropriate depending on the use case. We might want to partition a set of isolated cpus and still allow those cpus to load balance. With the addition of the spec.CPU.EnableIsolcpus boolean value, we can toggle on usage of isolcpus kernel args independently of the tuned partitioning.

## Changes

- Removed NonIsolated API option. This now defaults to Reserved
- Added EnableIsolcpus API option for toggling between using kernel isolcpus arg or not for the isolated cpuset. This option defaults to false.

### CODE REVIEW

File: deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml [hunk 1]
```diff
@@ -34,15 +34,23 @@ spec:
             cpu:
               description: CPU defines set of CPU related parameters.
               properties:
+                enableIsolcpus:
```
Comment (MarSik) [inline deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: Do we really need the boolean? What about just checking the isolated string is nil or empty?

Comment (davidvossel) [reply deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: the issue from what I could see is there's tuned partitioning that uses the Isolated CPUSet independently of whether or not we set the isolcpu kernel arg.

Comment (davidvossel) [reply deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: @MarSik the Isolated cpu set is used in two different configs.

1. In a tuned config to ste teh isolated_cores cpu set.
```
[scheduler]
isolated_cores="4-7"
```

2. In a MachineConfig to set the IRQBALANCE_BANNED_CPUS=11110000 bitmask for irqbalance.

```
echo "IRQBALANCE_BANNED_CPUS=11110000" >> ${INITRD_GENERATION_DIR}/etc/sysconfig/irqbalance
```

Since item [2] here lets us execute arbitrary bash logic on the actually host, the inverse of `Reserved` is possible. However for item [1], since the tuned profile is generated at the cluster level, the only way to compute the inverse of `Reserved` is to know how many cpus are on the matching hosts, which isn't something we can get from the node objects themselves by default.

So, basically the Isolated option is required because there isn't a reasonable way (that i'm aware of) to calculate the inverse of `Reserved` at the cluster level. Because of this, we need the `EnableIsolcpu` boolean to indicate if that kernel arg should be set or not independently of the `Isolated` cpu set

make sense?

**Comment by davidvossel** [comment]
updated to replace all references of nonIsolated with Reserved.

File: build/assets/scripts/pre-boot-tuning.sh [hunk 2]
```diff
@@ -74,8 +74,8 @@ else
 
     type getargs >/dev/null 2>&1 || . /lib/dracut-lib.sh
 
-    #cpumask="$(getargs non_iso_cpumask)"
-    cpumask='$non_iso_cpumask'
+    #cpumask="$(getargs reserved_cpumask)"
```
Comment (cynepco3hahue) [inline build/assets/scripts/pre-boot-tuning.sh]: I think you can remove this line

File: build/assets/tuned/openshift-node-real-time-kernel [hunk 3]
```diff
@@ -19,4 +19,4 @@ kernel.timer_migration = 0
 /sys/devices/system/machinecheck/machinecheck*/ignore_ce = 1   
 
 [scheduler]
-isolated_cores={{.IsolatedCpus}}
+{{if .IsolatedCpus}}isolated_cores={{.IsolatedCpus}}{{end}}
```
Comment (cynepco3hahue) [inline build/assets/tuned/openshift-node-real-time-kernel]: I think we should put the whole `[scheduler]` section under `if` condition

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go [hunk 4]
```diff
@@ -163,9 +168,9 @@ func getIgnitionConfig(assetsDir string, profile *performancev1alpha1.Performanc
 		})
 	}
 
-	nonIsolatedCpus := profile.Spec.CPU.NonIsolated
+	ReservedCpus := profile.Spec.CPU.Reserved
```
Comment (cynepco3hahue) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: ReservedCpus -> reservedCpus
true for all occurences

File: deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml [hunk 5]
```diff
@@ -34,15 +34,23 @@ spec:
...
+                  description: EnableIsolcpus results in the Isolated CPUset being
+                    listed in the isolcpu kernel arg on the host machines. When this
+                    option is set to "false", the Isolated cpus will be partitioned
+                    by tuned, but load balancing across the Isolated cpus will still
+                    be possible. When this option is set to "true", the Isolated CPU
```
Comment (yanirq) [inline deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: @davidvossel   In case we had the full cpu_partitioning profile support from NTO which also includes: no_balance_cores that disable the kernel load balancing in certain isolated CPUs, could we then say that this boolean will becomes redundant? (see [here](https://github.com/redhat-performance/tuned/blob/bfe0eb91949d4e305de5c22bec913fc8318b3a33/man/tuned-profiles-cpu-partitioning.7#L55))

File: deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml (hunk 1 above)
Comment (yanirq) [reply deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: @davidvossel is right regarding the difference of isolated_cores used in kargs and isolated_cores passed to the [tuned profie](https://github.com/openshift-kni/performance-addon-operators/blob/832b615823bd28807d0eef2748d808739f7b266e/build/assets/tuned/openshift-node-real-time-kernel#L22). 

> 2. In a MachineConfig to set the IRQBALANCE_BANNED_CPUS=11110000 bitmask for irqbalance.

the pre tuning script we have right now takes the reserved cpus list and inverts it to an isolated_cpumask.
On the other hand we [set](https://github.com/openshift-kni/performance-addon-operators/blob/832b615823bd28807d0eef2748d808739f7b266e/build/assets/scripts/pre-boot-tuning.sh#L37) CPUAffinity from the reserved cpu list.

Just as a reference, Tuned cpu-partitioning profile takes a list of isolated cpus and inverts them to non_isolated_cpus [list](https://github.com/redhat-performance/tuned/blob/bfe0eb91949d4e305de5c22bec913fc8318b3a33/profiles/cpu-partitioning/tuned.conf#L25) (or [mask](https://github.com/redhat-performance/tuned/blob/bfe0eb91949d4e305de5c22bec913fc8318b3a33/profiles/cpu-partitioning/tuned.conf#L28)) for setting CPUAffinity.  Supported py scripts for that can be found [here](https://github.com/redhat-performance/tuned/blob/master/tuned/profiles/functions/function_cpulist_invert.py).

File: deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml (hunk 5 above)
Comment (MarSik) [reply deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: Actually, this is pretty good reason to keep the argument. We can then always pass `isolated` to isolated_cores=<CPU\-LIST> and iff the boolean is set then set the same list to no_balance_cores as well (isolcpus for now).

File: build/assets/tuned/openshift-node-real-time-kernel (hunk 3 above)
Comment (MarSik) [reply build/assets/tuned/openshift-node-real-time-kernel]: And add the no_balance_cores={{. IsolatedCpus}} conditional of .enableIsolcpus (which we should rename)

Comment (MarSik) [reply build/assets/tuned/openshift-node-real-time-kernel]: @davidvossel This was basically my only comment now. Lets keep the boolean, just change the name to something like `balance_isolated_cores` that defaults to true. And add no_balance_cores={{. IsolatedCpus}} when the balancing is disabled.

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go (hunk 4 above)
Comment (MarSik) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: I think I fixed this one for David.

File: build/assets/scripts/pre-boot-tuning.sh (hunk 2 above)
Comment (davidvossel) [reply build/assets/scripts/pre-boot-tuning.sh]: fixed

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go (hunk 4 above)
Comment (davidvossel) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: yup, it's fixed, thanks @MarSik

File: build/assets/tuned/openshift-node-real-time-kernel (hunk 3 above)
Comment (davidvossel) [reply build/assets/tuned/openshift-node-real-time-kernel]: I fixed this by calling the boolean `balanceIsolated=[true|false]`. The no_balanace_cores is something I think we should follow up on in another pr as there's probably some discussion around how this option related to isolcpu.

**Comment by davidvossel** [comment]
@MarSik @yanirq @cynepco3hahue @fromanirh 

I updated the PR. The big change here is the "enableIsolcpu" boolean has changed to "balanceIsolated". The default for balanceIsolated=true. Functionally it works just like the enableIsolcpu=false.

**Comment by slintes** [comment]
As far as I understand this... ;)
/lgtm

Not sure if others want to review again, so putting a hold on it. Feel free to remove.
/hold

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig_test.go [hunk 6]
```diff
@@ -93,13 +93,37 @@ const expectedBootArguments = `
...
+
 var _ = Describe("Machine Config", func() {
 	It("should generate yaml with expected parameters", func() {
 		profile := testutils.NewPerformanceProfile("test")
 		profile.Spec.HugePages.Pages = append(profile.Spec.HugePages.Pages, performancev1alpha1.HugePage{
 			Count: 1024,
 			Size:  "2M",
 		})
+		f := false
+		profile.Spec.CPU.BalanceIsolated = &f
```
Comment (cynepco3hahue) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig_test.go]: please use `pointer` module for it

```go
import "k8s.io/utils/pointer"
...
profile.Spec.CPU.BalanceIsolated = pointer.BoolPtr(false)
```

File: deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml [hunk 7]
```diff
@@ -34,15 +34,22 @@ spec:
...
+                balanceIsolated:
+                  description: BalanceIsolated toggles whether or not the Isolated
+                    CPU set is eligible for load balancing work loads. When this option
+                    is set to "false", the Isolated CPU set will be static, meaning
+                    workloads have to explicitly assign each thread to a specific
+                    cpu in order to work across multiple CPUs. Setting this to "true"
+                    allows workloads to be balanced across CPUs. Setting this to "false"
+                    offers the most predictable performance for guaranteed workloads,
+                    but it offloads the complexity of cpu load balancing to the application.
+                    Defaults to "true"
```
Comment (cynepco3hahue) [inline deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: Should it default to false, to preserve the previous behavior?

Comment (MarSik) [reply deploy/crds/performance.openshift.io_performanceprofiles_crd.yaml]: Nobody uses it yet and the previous behavior was actually one of the issues. I would prefer enabled balancing by default myself.

**Comment by MarSik** [comment]
> I still unsure if we need `balanceIsolated` parameter needed at all, but I will not block this PR because of it.

Yep, we do. Tuned profiles support that option as well.

**Comment by cynepco3hahue** [comment]
> Yep, we do. Tuned profiles support that option as well.

@MarSik Can you please give the use case, because, for the DPDK workload you want to have isolated CPU, and you do not want that some process will be balanced on top of it.
